traffic_ops/traffic_ops_golang/trafficvault/backends/postgres/postgres.go

```diff
@@ -100,24 +101,149 @@ func (p *Postgres) commitTransaction(tx *sqlx.Tx, ctx context.Context, cancelFun
+		return err
+	}
+	res, err := tvTx.Exec("INSERT INTO sslkey (deliveryservice, data, cdn, version) VALUES ($1, $2, $3, $4)", key.DeliveryService, keyJSON, key.CDN, key.Version)
+	if err != nil {
+		e := checkErrWithContext("Traffic Vault PostgreSQL: executing INSERT SSL Key query", err, ctx.Err())
+		return e
+	}
+	if rowsAffected, err := res.RowsAffected(); err != nil {
+		return err
+	} else if rowsAffected == 0 {
+		return errors.New("SSL Key: no keys were inserted")
+	}
+	return nil
 }
 
+// DeleteDeliveryServiceSSLKeys removes the SSL keys of the given version (or latest
+// if version is empty) for the delivery service identified by the given xmlID.
 func (p *Postgres) DeleteDeliveryServiceSSLKeys(xmlID string, version string, tx *sql.Tx, ctx context.Context) error {
-	return notImplementedErr
+	tvTx, dbCtx, cancelFunc, err := p.beginTransaction(ctx)
+	if err != nil {
+		return err
+	}
+	defer p.commitTransaction(tvTx, dbCtx, cancelFunc)
+	query := "DELETE FROM sslkey WHERE deliveryservice=$1"
+	if version != "" {
+		query += " AND version=$2"
+		_, err = tvTx.Exec(query, xmlID, version)
+	} else {
+		_, err = tvTx.Exec(query, xmlID)
+	}
+	if err != nil {
+		e := checkErrWithContext("Traffic Vault PostgreSQL: executing DELETE SSL Key query", err, ctx.Err())
+		return e
+	}
+	return nil
 }
 
+// DeleteOldDeliveryServiceSSLKeys takes a set of existingXMLIDs as input and will remove
+// all SSL keys for delivery services in the CDN identified by the given cdnName that
+// do not contain an xmlID in the given set of existingXMLIDs. This method is called
+// during a snapshot operation in order to delete SSL keys for delivery services that
+// no longer exist.
 func (p *Postgres) DeleteOldDeliveryServiceSSLKeys(existingXMLIDs map[string]struct{}, cdnName string, tx *sql.Tx, ctx context.Context) error {
-	return notImplementedErr
+	var keys []string
+	tvTx, dbCtx, cancelFunc, err := p.beginTransaction(ctx)
+	if err != nil {
+		return err
+	}
+	defer p.commitTransaction(tvTx, dbCtx, cancelFunc)
+
+	if len(existingXMLIDs) == 0 {
+		keys = append(keys, "")
+	}
+	for k, _ := range existingXMLIDs {
+		keys = append(keys, k)
+	}
+	_, err = tvTx.Exec("DELETE FROM sslkey WHERE cdn=$1 AND deliveryservice <> ALL ($2)", cdnName, pq.Array(keys))
+	if err != nil {
+		e := checkErrWithContext("Traffic Vault PostgreSQL: executing DELETE OLD SSL Key query", err, ctx.Err())
+		return e
+	}
+	return nil
 }
 
+// GetCDNSSLKeys retrieves all the SSL keys for delivery services in the CDN identified
+// by the given cdnName.
 func (p *Postgres) GetCDNSSLKeys(cdnName string, tx *sql.Tx, ctx context.Context) ([]tc.CDNSSLKey, error) {
-	return nil, notImplementedErr
+	var keys []tc.CDNSSLKey
+	var key tc.CDNSSLKey
+
+	tvTx, dbCtx, cancelFunc, err := p.beginTransaction(ctx)
+	if err != nil {
+		return keys, err
+	}
+	defer p.commitTransaction(tvTx, dbCtx, cancelFunc)
+
+	rows, err := tvTx.Query("SELECT data from sslkey WHERE cdn=$1", cdnName)
```

Comment:
we will also want to filter this query to only get the latest keys for each ds

Reply:
Yes, this query should only select keys `WHERE cdn=$1 AND version = 'latest'`